samza-azure/src/main/java/org/apache/samza/AzureJobCoordinator.java

```diff
@@ -0,0 +1,402 @@
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.samza.config.ApplicationConfig;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.ConfigException;
+import org.apache.samza.coordinator.JobCoordinator;
+import org.apache.samza.coordinator.JobCoordinatorListener;
+import org.apache.samza.coordinator.JobModelManager;
+import org.apache.samza.coordinator.LeaderElectorListener;
+import org.apache.samza.job.model.JobModel;
+import org.apache.samza.runtime.ProcessorIdGenerator;
+import org.apache.samza.scheduler.HeartbeatScheduler;
+import org.apache.samza.scheduler.JMVersionUpgradeScheduler;
+import org.apache.samza.scheduler.LeaderBarrierCompleteScheduler;
+import org.apache.samza.scheduler.LeaderLivenessCheckScheduler;
+import org.apache.samza.scheduler.LivenessCheckScheduler;
+import org.apache.samza.scheduler.RenewLeaseScheduler;
+import org.apache.samza.scheduler.SchedulerStateChangeListener;
+import org.apache.samza.system.StreamMetadataCache;
+import org.apache.samza.util.ClassLoaderHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Class that provides coordination mechanism for Samza standalone in Azure.
+ * Handles processor lifecycle through Azure blob and table storage. Orchestrates leader election.
+ * The leader job coordinator generates partition mapping, writes shared data to the blob and manages rebalancing.
+ */
+public class AzureJobCoordinator implements JobCoordinator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AzureJobCoordinator.class);
+  private static final int METADATA_CACHE_TTL_MS = 5000;
+  private static final String INITIAL_STATE = "0";
+  private final AzureLeaderElector azureLeaderElector;
+  private final BlobUtils leaderBlob;
+  private final TableUtils table;
+  private final Config config;
+  private final String processorId;
+  private final AzureClient client;
+  private final AtomicReference<String> currentJMVersion;
+  private final AtomicBoolean versionUpgradeDetected;
+  private final HeartbeatScheduler heartbeat;
+  private final JMVersionUpgradeScheduler versionUpgrade;
+  private final LeaderLivenessCheckScheduler leaderAlive;
+  private final LivenessCheckScheduler liveness;
+  private RenewLeaseScheduler renewLease;
+  private LeaderBarrierCompleteScheduler barrierState;
+  private ScheduledFuture leaderBarrierSF;
+  private StreamMetadataCache streamMetadataCache = null;
+  private JobCoordinatorListener coordinatorListener = null;
+  private JobModel jobModel = null;
+
+  /**
+   * Creates an instance of Azure job coordinator, along with references to Azure leader elector, Azure Blob and Azure Table.
+   * @param config User defined config
+   */
+  public AzureJobCoordinator(Config config) {
+    this.config = config;
+    processorId = createProcessorId(config);
+    currentJMVersion = new AtomicReference<>(INITIAL_STATE);
```

Comment:
Why is the initial value for `currentJMVersion` 0 ? Can you explain how you are using this variable?